File: f1-prediction-app/frontend/src/utils/raceCalendar.js
export const getRaceCalendar = async () => {
  try {
    const response = await fetch('http://127.0.0.1:5000/api/race-calendar');
    if (!response.ok) {
      throw new Error('Failed to fetch race calendar');
    }
    return await response.json();
  } catch (error) {
    console.error('Error fetching race calendar:', error);
    return []; // Fallback to an empty array or a hardcoded calendar
  }
};

export const getNextRace = async () => {
  const raceCalendar = await getRaceCalendar();
  if (!raceCalendar || raceCalendar.length === 0) {
    return null;
  }
  const now = new Date();
  return raceCalendar.find(race => new Date(race.date) > now) || raceCalendar[0];
};

export const getLastRace = async () => {
  const raceCalendar = await getRaceCalendar();
  if (!raceCalendar || raceCalendar.length === 0) {
    return null;
  }
  const now = new Date();
  return [...raceCalendar]
    .reverse()
    .find(race => new Date(race.date) < now) || raceCalendar[raceCalendar.length - 1];
};

File: f1-prediction-app/backend/services/race_calendar.py
import fastf1
import pandas as pd
from datetime import datetime
import logging

class RaceCalendarService:
    def __init__(self):
        self.current_year = datetime.now().year

    def get_race_calendar(self):
        """Fetch the current season's race calendar."""
        try:
            schedule = fastf1.get_event_schedule(self.current_year)
            race_calendar = []
            for _, event in schedule.iterrows():
                race_calendar.append({
                    'name': event['EventName'],
                    'date': event['EventDate'].strftime('%Y-%m-%d'),
                    'time': event['Session5Date'].strftime('%H:%M') if pd.notna(event['Session5Date']) else 'TBD',
                    'sprint': 'Sprint' in event['EventFormat']
                })
            return race_calendar
        except Exception as e:
            logging.error(f"Error fetching race calendar: {str(e)}")
            return None